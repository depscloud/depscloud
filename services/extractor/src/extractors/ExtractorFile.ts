import cheerio = require("cheerio");
import YAML = require("js-yaml");
import TOML = require("toml");

export default class ExtractorFile {
    private readonly body: string;

    constructor(body: string) {
        this.body = body;
    }

    public raw(): string {
        return this.body;
    }

    public json(): any {
        return JSON.parse(this.body);
    }

    public toml(): any {
        return TOML.parse(this.body);
    }

    public yaml(): any {
        return YAML.safeLoad(this.body);
    }

    public xml(): cheerio.Cheerio {
        return cheerio.load(this.body).root();
    }
}
