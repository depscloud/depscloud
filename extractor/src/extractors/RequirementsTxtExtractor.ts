import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import MatchConfig from "../matcher/MatchConfig";
import {ManifestFile} from "@depscloud/api/v1beta";
import Languages from "./Languages";
import inferPythonName from "./utils/inferPythonName";
import {ManifestDependency} from "@depscloud/api/depscloud/api/v1beta/manifest";

const operators = [
    "~", "=", "!", ">", "<"
];

export default class RequirementsTxtExtractor implements Extractor {
    public async extract(url: string, files: { [p: string]: ExtractorFile }): Promise<ManifestFile> {
        const dependencies: Array<ManifestDependency> = [];

        const contents = files["requirements.txt"].raw();
        const lines = contents.split("\n");

        lines.forEach((line) => {
            const commentIndex = line.indexOf("#");
            if (commentIndex > -1) {
                line = line.substring(0, commentIndex);
            }

            line = line.trim()

            if (!line || line[0] === "-") {
                return
            }

            let l = line.length;
            for (const operator of operators) {
                const i = line.indexOf(operator);
                if (i > -1) {
                    l = Math.min(l, i);
                }
            }

            const name = line.substring(0, l).trim();
            const versionConstraint = line.substring(l).trim();

            dependencies.push({
                name,
                versionConstraint: (versionConstraint || "*"),
                scopes: [""],
            });
        });

        return {
            language: Languages.PYTHON,
            system: "pip",
            sourceUrl: "",
            name: inferPythonName(url),
            version: "*",
            dependencies,
        };
    }

    public matchConfig(): MatchConfig {
        return {
            includes: [ "**/requirements.txt" ],
            excludes: [],
        };
    }

    public requires(): string[] {
        return [ "requirements.txt" ];
    }
}
