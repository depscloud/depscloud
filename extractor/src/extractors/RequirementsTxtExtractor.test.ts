import {readFile} from "fs";
import {promisify} from "util";
import ExtractorFile from "./ExtractorFile";
import RequirementsTxtExtractor from "./RequirementsTxtExtractor";

const readFileAsync = promisify(readFile);

describe("RequirementsTxtExtractor", () => {
    test("fullParse", async () => {
        const filePath = require.resolve("./testdata/requirements.txt");
        const buffer = await readFileAsync(filePath);
        const content = buffer.toString();

        const parser = new RequirementsTxtExtractor();

        const url = "git@example.com:common/python-web.git";
        const actual = await parser.extract(url, { "requirements.txt": new ExtractorFile(content) });

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
