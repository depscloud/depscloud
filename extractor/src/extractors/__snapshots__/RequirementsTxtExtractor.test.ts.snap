// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequirementsTxtExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "nose",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "nose-cov",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "beautifulsoup4",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "docopt",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "== 0.6.1",
    },
    Object {
      "name": "keyring",
      "scopes": Array [
        "",
      ],
      "versionConstraint": ">= 4.1.1",
    },
    Object {
      "name": "coverage",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "!= 3.5",
    },
    Object {
      "name": "Mopidy-Dirble",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "~= 1.1",
    },
    Object {
      "name": "./downloads/numpy-1.9.2-cp34-none-win32.whl",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "http://wxpython.org/Phoenix/snapshot-builds/wxPython_Phoenix-3.0.3.dev1820+49a8884-cp34-none-win_amd64.whl",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "rejected",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "green",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "*",
    },
  ],
  "language": "python",
  "name": "python-web",
  "sourceUrl": "",
  "system": "pip",
  "version": "*",
}
`;

exports[`RequirementsTxtExtractor fullParse 2`] = `
"{
  \\"language\\": \\"python\\",
  \\"system\\": \\"pip\\",
  \\"sourceUrl\\": \\"\\",
  \\"name\\": \\"python-web\\",
  \\"version\\": \\"*\\",
  \\"dependencies\\": [
    {
      \\"name\\": \\"nose\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"nose-cov\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"beautifulsoup4\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"docopt\\",
      \\"versionConstraint\\": \\"== 0.6.1\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"keyring\\",
      \\"versionConstraint\\": \\">= 4.1.1\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"coverage\\",
      \\"versionConstraint\\": \\"!= 3.5\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"Mopidy-Dirble\\",
      \\"versionConstraint\\": \\"~= 1.1\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"./downloads/numpy-1.9.2-cp34-none-win32.whl\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"http://wxpython.org/Phoenix/snapshot-builds/wxPython_Phoenix-3.0.3.dev1820+49a8884-cp34-none-win_amd64.whl\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"rejected\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"name\\": \\"green\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"\\"
      ]
    }
  ]
}"
`;
